Add stress test to System.Threading.Channels

tarekgh opened the pull request.

tarekgh commented:
@stephentoub could you please have a look?

stephentoub commented on src/System.Threading.Channels/tests/Stress.cs:
```diff
@@ -0,0 +1,423 @@
+        private static readonly UnbufferedChannelOptions UnbufferedSynchronousSingleWriterOptions =
+                                    new UnbufferedChannelOptions() { AllowSynchronousContinuations = true, SingleWriter = true };
+        private static readonly UnbufferedChannelOptions UnbufferedSynchronousSingleReaderWriterOptions =
+                                    new UnbufferedChannelOptions() { AllowSynchronousContinuations = true, SingleReader = true, SingleWriter = true };
```
Do we need to define all these fields?  They look like they're each only used in a single yield; seems like it'd be easier to understand the test data if everything to do with each yield was colocated in that yield.  You could also change the test data to avoid needing to refer to each twice, e.g.
```C#
public static IEnumerable<object[]> TestData()
{
    Func<object, Channel<int>> unbounded = o => Channel.CreateUnbounded<int>((UnboundedChannelOptions)o);
    yield return new object[] { unbounded, new UnboundedChannelOptions() { SingleReader = true } };
    ...
}
etc.  That then also lets you more easily use loops to avoid needing to explicitly write out every option, e.g.
```C#
foreach (bool singleReader in new[] { true, false })
{
    foreach (bool singleWriter in new[] { true, false })
    {
        yield return new object[] { unbounded, new UnboundedChannelOptions { SingleReader = singleReader, SingleWriter = singleWriter } };
    }
}
```
etc.

tarekgh replied:
ok will do that.

stephentoub commented on src/System.Threading.Channels/tests/Stress.cs:
```diff
@@ -0,0 +1,423 @@
+        const int MAX_NUMBER_TO_WRITE_TO_CHANNEL = 1_000_000;
+        private static readonly int MaxTaskCounts = Math.Max(2, Environment.ProcessorCount);
+
+        [OuterLoop]
```
How long does this test take to run?  If it takes a long time, even in outerloop that could be problematic.  And if it doesn't take a long time, what is it stressing?

tarekgh replied:
No, it doesn't take that long. on my machine, it takes less than 2 minutes. it should be faster after we remove the unbuffered channel too. so I think keeping it in the outerloop is ok.

stephentoub replied:
> it takes less than 2 minutes

I think that's way too long, even for an Outerloop test.

@danmosemsft, @karelz, what's the recommendation for stress tests in corefx?

tarekgh replied:
I just measured it (for the 1,000,000 iterations) and it took 1 minute.

stephentoub commented on src/System.Threading.Channels/tests/Stress.cs:
```diff
@@ -0,0 +1,423 @@
+            yield return new object[] { Channel.CreateUnbuffered<int>(UnbufferedSynchronousOptions), UnbufferedSynchronousOptions };
+        }
+
+        const int MAX_NUMBER_TO_WRITE_TO_CHANNEL = 1_000_000;
```
Nit: PascalCasing for consts

tarekgh replied:
I'll fix that

stephentoub commented on src/System.Threading.Channels/tests/Stress.cs:
```diff
@@ -0,0 +1,423 @@
+
+            for (int i=0; i < readerTasksCount; i++)
+            {
+                taskList.Add(new Task(() => {
```
Any specific reason this is using Task's ctor and then calling Start, rather than using Task.Run?

tarekgh replied:
mo specific reason. I'll change it to Task.Run.

stephentoub commented on src/System.Threading.Channels/tests/Stress.cs:
```diff
@@ -0,0 +1,423 @@
+                        {
+                            writer.WriteAsync(incrementedValue).GetAwaiter().GetResult();
+                        }
+                        Interlocked.Increment(ref wroteCount);
```
There is a simpler way to coordinate the writers.  You just need to have the last one call writer.Complete, e.g.
```C#
int remainingWriters = writerTasksCount;
...
taskList.Add(Task.Run(async delegate
{
    int i = 0;
    while (i < MaxNumberToWriteToChannel)
    {
        await writer.WriteAsync(i);
        i = Interlocked.Increment(ref numberToWriteToQueue);
    }
    if (Interlocked.Decrement(ref remainingWriters) == 0) writer.Complete();
});
```

tarekgh replied:
ok. will do that

stephentoub commented on src/System.Threading.Channels/tests/Stress.cs:
```diff
@@ -0,0 +1,423 @@
+
+                        if (!writer.TryWrite(incrementedValue))
+                        {
+                            writer.WriteAsync(incrementedValue).GetAwaiter().GetResult();
```
What's the goal of this pattern?  Note that for the unbounded channel, WriteAsync will never be used as TryWrite will always succeed.  So if there's a desire in this test to test WriteAsync, that won't happen in such cases.

tarekgh replied:
I am trying to write the code that can work with all channel even without the knowledge what type of channel I am using. if you have better pattern suggestion that works with all channels, I am ok to use. note that, I wrote that to work even with the unbuffered channels.

stephentoub commented on src/System.Threading.Channels/tests/Stress.cs:
```diff
@@ -0,0 +1,423 @@
+        [OuterLoop]
+        [Theory]
+        [MemberData(nameof(TestData))]
+        public void RunInStressMode(Channel<int> channel, ChannelOptions options)
```
What are you trying to stress here specifically?

tarekgh replied:
I am trying to stress simultaneous read and write operation on the channels with different channels options.

tarekgh commented on src/System.Threading.Channels/tests/Stress.cs:
```diff
@@ -0,0 +1,423 @@
+                        {
+                            if (!reader.TryRead(out int value))
+                            {
+                                value = reader.ReadAsync().GetAwaiter().GetResult();
```
I noticed when I was using wait, the task was signaled (in Task.WaitAll) even the whole reading/writing is not executed yet. this is why I preferred the blocking. I'll take a look again at this.

stephentoub replied:
That's because you were using Task's ctor, which used in this way would take an Action, and that Action would return to the Task ctor on the first await.
https://blogs.msdn.microsoft.com/pfxteam/2012/02/08/potential-pitfalls-to-avoid-when-passing-around-async-lambdas/
That won't be an issue when you switch to use Task.Run.

tarekgh commented:
@stephentoub  I have addressed all feedback but note that I removed the unbuffered changes here as we are going to get rid of unbuffed channel completely. please let me know if you have any further comments and thanks for the feedback.

tarekgh commented:
#24823

tarekgh commented:
test OSX x64 Debug Build

stephentoub commented on src/System.Threading.Channels/tests/Stress.cs:
```diff
@@ -0,0 +1,144 @@
+        [OuterLoop]
+        [Theory]
+        [MemberData(nameof(TestData))]
+        public void RunInStressMode(Func<ChannelOptions, Channel<int>> channelCreator, ChannelOptions options)
```
Is this meant to be the only stress test?  I would have expected different tests for different things.  For example, this doesn't test WaitForRead/WriteAsync, it doesn't test TryWrite, and it likely doesn't exercise various patterns of member X being used concurrently with member Y.  It also doesn't stress completion, how it interacts with concurrent attempts to read/write, etc.

The test should be named for the thing that it's trying to stress, and ideally there would be tests for all interesting aspects.

That then contributes to my concern about this being [Outerloop].  Even if this test takes 1 minute (which I still think is too long for Outerloop... imagine if every project had an additional 1 minute test, that could add hours to a corefx run), if each such test took a minute we'd be at much more than that.

This test seems like a reasonable stressing of WriteAsync + TryRead/ReadAsync, although even then I'm not sure it's really going to stress WriteAsync+ReadAsync in a way that might be desired, but a lot of the reads will be satisfied by TryRead.

tarekgh replied:
- Currently we don't have the infrastructure in corefx for stress tests. I already talked before with @weshaggard and he suggested to keep such test as outerloop. if you have a better suggestion, please let me know and I'll be happy to do it.

- I'll add another test to ensure exercising WriteAsync and ReadAsync (with using WaitForRead/Write.

stephentoub replied:
> Currently we don't have the infrastructure in corefx for stress tests. I already talked before with @weshaggard and he suggested to keep such test as outerloop

@karelz, is this how we want to handle stress tests?

stephentoub commented on src/System.Threading.Channels/tests/Stress.cs:
```diff
@@ -0,0 +1,144 @@
+                        }
+                        catch (ChannelClosedException)
+                        {
+                            break;
```
You can move the try/catch to be outside of the while loop.  Then this break isn't necessary, and we're also not entering/leaving the try on each iteration.

tarekgh replied:
I'll do that although the intention of this test is to optimize the execution. it is just a stress test which should be doing some not desired patterns too.

stephentoub commented on src/System.Threading.Channels/tests/Stress.cs:
```diff
@@ -0,0 +1,144 @@
+
+            Task.WaitAll(taskList.ToArray());
+
+            Assert.True(readCount <= MaxNumberToWriteToChannel);
```
Why wouldn't they be equal?

stephentoub replied:
Ah, nevermind, I see the shouldReadAllWrittenValues below.  In that case I'd suggest a small tweak:
```C#
if (shouldReadAllWrittenValues)
{
    Assert.Equal(MaxNumberToWriteToChannel, readCount);
}
else
{
    Assert.InRange(readCount, 0, MaxNumberToWriteToChannel);
}
```

tarekgh replied:
ok, I'll add this.

stephentoub commented on src/System.Threading.Channels/tests/Stress.cs:
```diff
@@ -0,0 +1,144 @@
+            //                  (boundedOptions == null ? "" : $"Mode: {boundedOptions.FullMode}, Capacity: {boundedOptions.Capacity},") +
+            //                  $"read: {readCount}, " +
+            //                  $"Max: {MaxNumberToWriteToChannel}";
+            // Console.WriteLine(message);
```
Delete this comment?

tarekgh replied:
I was keeping it because it is handy to enable for more verbose output when trying to diagnose the execution. but I'll delete it anyway.

tarekgh commented:
> Looking better

Yes, it is much better. thanks for all your comments.

tarekgh commented:
@stephentoub 

- I have added the different reading and writing pattern to the tests. please let me know if you would like to see more patterns to test.
- I have talked to @danmosemsft regarding having the test as OuterLoop, he mentioned this is fine as long the whole library test runs under the timeout limit which is currently 20 minutes. After my current change with supporting different reading and writing patterns, the test is running for 5 minutes (on my machine) with using 400,000 iterations for every channel with different options. So I believe we should be fine here. 

feel free to send more suggestions or comments. I am seeing the test is getting much better and valuable.

tarekgh commented:
@stephentoub PTAL. thanks.

tarekgh commented:
#26808

stephentoub commented on src/System.Threading.Channels/tests/Stress.cs:
```diff
@@ -0,0 +1,212 @@
+
+        public static IEnumerable<object[]> TestData()
+        {
+            foreach (var readDelegate in new Read[] { ReadSynchronous, ReadASynchronous, ReadSyncAndASync} )
```
Nit: ReadASynchronous => ReadAsynchronous
Nit: ReadSyncAndASync => ReadSyncAndAsync

stephentoub commented on src/System.Threading.Channels/tests/Stress.cs:
```diff
@@ -0,0 +1,212 @@
+    public class StressTests
+    {
+        public delegate Task<bool> Read(ChannelReader<int> reader);
+        public delegate Task Write(ChannelWriter<int> writer, int value);
```
Nit: ReadFunc, WriteFunc, or something like that

stephentoub replied:
But you can also just use `Func<...>`

stephentoub commented on src/System.Threading.Channels/tests/Stress.cs:
```diff
@@ -0,0 +1,212 @@
+
+            foreach (var readDelegate in new Read[] { ReadSynchronous, ReadASynchronous, ReadSyncAndASync} )
+            foreach (var writeDelegate in new Write[] { WriteSynchronous, WriteASynchronous, WriteSyncAndASync} )
+            for (BoundedChannelFullMode bco = BoundedChannelFullMode.Wait;  bco <= BoundedChannelFullMode.DropWrite; bco++)
```
Could be:
```C#
foreach (BoundedChannelFullMode bco in Enum.GetValues(typeof(BoundedChannelFullMode)))
```

stephentoub commented on src/System.Threading.Channels/tests/Stress.cs:
```diff
@@ -0,0 +1,212 @@
+
+        const int MaxNumberToWriteToChannel = 400_000;
+        private static readonly int MaxTaskCounts = Math.Max(2, Environment.ProcessorCount);
+
```
Now that you've added the TestEnvironment thing, let's use it here.  Thanks for doing that.

tarekgh commented:
Thanks @stephentoub for all comments and your review.
